src/Saturn/Channels.fs

```diff
@@ -0,0 +1,139 @@
+namespace Saturn
+
+open System
+open System.Threading
+open System.Threading.Tasks
+open System.Net.WebSockets
+open Microsoft.AspNetCore.Http
+open FSharp.Control.Tasks.V2
+open System.Collections.Generic
+open System.Collections.Concurrent
+open Giraffe.Serialization.Json
+
+module Channels =
+    let internal sockets = Dictionary<string, ConcurrentDictionary<string, WebSocket>>()
+
+    type Message = {Topic: string; Ref: string; Payload: obj}
+
+    type JoinResult =
+        | Ok
+        | Rejected of reason: string
+
+    type IChannel =
+        abstract member Join: HttpContext -> Task<JoinResult>
+        abstract member HandleMessage: HttpContext * WebSocketReceiveResult * Message -> Task<unit>
+        abstract member Terminate: HttpContext -> Task<unit>
+
+    type SocketMiddleware(next : RequestDelegate, serializer: IJsonSerializer, path: string, channel: IChannel) =
+        do sockets.Add(path, ConcurrentDictionary())
+
+        member __.Invoke(ctx : HttpContext) =
+            task {
+                if ctx.Request.Path = PathString(path) then
+                    match ctx.WebSockets.IsWebSocketRequest with
+                    | true ->
+                        let! joinResult = channel.Join ctx
+                        match joinResult with
+                        | Ok ->
+                            let! webSocket = ctx.WebSockets.AcceptWebSocketAsync()
+                            let guid = Guid.NewGuid().ToString()
+                            sockets.[path].AddOrUpdate (guid, webSocket, fun _ _ -> webSocket) |> ignore
+
+                            let buffer : byte [] = Array.zeroCreate 4096
```

Comment:
You shouldn't rely on 4096 being the message size.  `echo` will have a `EndOfMessage` bool telling you if that message frame as more to read.  See https://github.com/TheAngryByrd/FSharp.Control.WebSockets/blob/master/src/FSharp.Control.Websockets.TPL/FSharp.Control.Websockets.TPL.fs#L54-L77 for an example you can steal.